### DISCUSSION TITLE
Question regarding Immer in combination with no store actions.

### DISCUSSION THREAD

**Discussion by JKannekens** [pmndrs/zustand#2312] [Q&A] [answered]
I want to use immer in combination with no store actions, because I want to have all the actions in separate files. However when I add the immer middleware. It doesn't get the correct type for the state. I am not sure if I am doing this correctly, but if I would move the State type from create to Immer it works.

Example:
``` typescript
interface Todo {
    id: string;
    title: string;
    done: boolean;
}

type State = {
    todos: Record<string, Todo>;
};

const useTodoStore = create<State>()(
    immer(() => ({
        todos: {
            "82471c5f-4207-4b1d-abcb-b98547e01a3e": {
                id: "82471c5f-4207-4b1d-abcb-b98547e01a3e",
                title: "Learn Zustand",
                done: false,
            },
            "354ee16c-bfdd-44d3-afa9-e93679bda367": {
                id: "354ee16c-bfdd-44d3-afa9-e93679bda367",
                title: "Learn Jotai",
                done: false,
            },
            "771c85c5-46ea-4a11-8fed-36cc2c7be344": {
                id: "771c85c5-46ea-4a11-8fed-36cc2c7be344",
                title: "Learn Valtio",
                done: false,
            },
            "363a4bac-083f-47f7-a0a2-aeeee153a99c": {
                id: "363a4bac-083f-47f7-a0a2-aeeee153a99c",
                title: "Learn Signals",
                done: false,
            },
        },
    }))
);
```

Thank you in advance!

**Comment by dbritto-dev** [comment]
@JKannekens did you install immer?

**Reply by JKannekens** [reply]
Yes I did, I made an example in code sandbox: [Link to codesandbox](https://codesandbox.io/p/sandbox/immer-zustand-39l6qw?file=%2Fsrc%2FApp.tsx%3A45%2C16&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clrus1t7m00063d5ut6llj17l%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clrus1t7m00023d5u4pwvyp02%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clrus1t7m00033d5uw116eprd%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clrus1t7m00053d5uaiq5twvq%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clrus1t7m00023d5u4pwvyp02%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrus1t7m00013d5uwalgie1x%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpublic%252Findex.html%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clrus976s00023d5uzbc4tox5%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A45%252C%2522startColumn%2522%253A16%252C%2522endLineNumber%2522%253A45%252C%2522endColumn%2522%253A16%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clrus1t7m00023d5u4pwvyp02%2522%252C%2522activeTabId%2522%253A%2522clrus976s00023d5uzbc4tox5%2522%257D%252C%2522clrus1t7m00053d5uaiq5twvq%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrus1t7m00043d5ukpw2ml3x%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clrus1t7m00053d5uaiq5twvq%2522%252C%2522activeTabId%2522%253A%2522clrus1t7m00043d5ukpw2ml3x%2522%257D%252C%2522clrus1t7m00033d5uw116eprd%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clrus1t7m00033d5uw116eprd%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Reply by dbritto-dev** [reply, marked as answer]
@JKannekens here you go -> https://codesandbox.io/p/sandbox/immer-zustand-forked-qhtm3k?file=%2Fsrc%2FApp.tsx%3A15%2C18 I bet in your case is inferring the type from the object itself but there's a workaround that you can use to deal with that

**Reply by JKannekens** [reply]
Oh wow, didn't think about that. Thanks for the fast response!

**Reply by naseervn** [reply]
> @JKannekens here you go -> https://codesandbox.io/p/sandbox/immer-zustand-forked-qhtm3k?file=%2Fsrc%2FApp.tsx%3A15%2C18 I bet in your case is inferring the type from the object itself but there's a workaround that you can use to deal with that

this link is not active. can I please get a active link, a bit curious about the answer

**Reply by jeremiahfallin** [reply]
Also looking for an example of this.

**Reply by dbritto-dev** [reply]
In a nutshell is this:

```ts
interface Todo {
    id: string;
    title: string;
    done: boolean;
}

type State = {
    todos: Record<string, Todo>;
};

const initialTodos: Record<string, Todo> = {
      "82471c5f-4207-4b1d-abcb-b98547e01a3e": {
          id: "82471c5f-4207-4b1d-abcb-b98547e01a3e",
          title: "Learn Zustand",
          done: false,
      },
      "354ee16c-bfdd-44d3-afa9-e93679bda367": {
          id: "354ee16c-bfdd-44d3-afa9-e93679bda367",
          title: "Learn Jotai",
          done: false,
      },
      "771c85c5-46ea-4a11-8fed-36cc2c7be344": {
          id: "771c85c5-46ea-4a11-8fed-36cc2c7be344",
          title: "Learn Valtio",
          done: false,
      },
      "363a4bac-083f-47f7-a0a2-aeeee153a99c": {
          id: "363a4bac-083f-47f7-a0a2-aeeee153a99c",
          title: "Learn Signals",
          done: false,
      },
  }

const useTodoStore = create<State>()(
    immer(() => ({ todos: initialTodos, }))
);
```
